### PR TITLE
Toggle between basic, full, and linguistic paradigms

### PR DESCRIPTION
**Pull request by Madoshakalaka**
Now the paradigms can be toggled by a button.

Also visitable from the link with query param `paraidmg-size`

### CODE REVIEW

File: CreeDictionary/API/models.py [hunk 1]
```diff
@@ -230,12 +230,15 @@ def homograph_disambiguator(self) -> Optional[str]:
     def full_word_class(self) -> Optional[WordClass]:
         return fst_analysis_parser.extract_word_class(self.analysis)
 
-    @property
-    def paradigm(self) -> List[Layout]:
-        # todo: allow paradigm size other then ParadigmSize.BASIC
+    def get_paradigm_layouts(
+        self, size: ParadigmSize = ParadigmSize.BASIC
```
Comment (eddieantonio) [inline CreeDictionary/API/models.py:234]: Nice backwards compatibility!

File: CreeDictionary/CreeDictionary/templates/CreeDictionary/base.html [hunk 2]
```diff
@@ -6,92 +6,112 @@
...
-    <template id="template:recording-item">
-      <li class="recordings-list__item" data-cy="recordings-list__item">
-        <button class="audio-snippet"><slot name="speaker-name"></slot>, <slot name="speaker-dialect"></slot></button>
-    </template>
+  {# The speaker button template (multiple recordings) #}
+  <template id="template:recording-item">
+    <li class="recordings-list__item" data-cy="recordings-list__item">
+      <button class="audio-snippet">
+        <slot name="speaker-name"></slot>
+        ,
```
Comment (eddieantonio) [inline CreeDictionary/CreeDictionary/templates/CreeDictionary/base.html:49]: IntelliJ is not smart about this :(

There should be a comma directly after the `<slot>` or else there will be unnecessary whitespace

```suggestion
        <slot name="speaker-name"></slot>,
```

File: CreeDictionary/CreeDictionary/templates/CreeDictionary/base.html [hunk 3]
```diff
@@ -6,92 +6,112 @@
...
+  {% if lemma_id and paradigm_size%}
+    {% comment %}
+    meaning we are on lemma details/paradigm page. This variable can be utilized by javascript to dynamically request
+    other things (e.g. different sized paradigms) about the lemma
+    {% endcomment %}
+    <script>
+      var lemmaId = {{ lemma_id }}
+      var paradigmSize = '{{ paradigm_size.value }}'
+    </script>
+  {% endif %}
```
Comment (eddieantonio) [inline CreeDictionary/CreeDictionary/templates/CreeDictionary/base.html:25]: Consider using  `json_script` instead: https://docs.djangoproject.com/en/3.0/ref/templates/builtins/#json-script

File: CreeDictionary/CreeDictionary/views.py [hunk 4]
```diff
@@ -86,18 +103,55 @@ def search_results(request, query_string: str):  # pragma: no cover
     )
 
 
-def lemma_details_internal(request, lemma_id: int):  # pragma: no cover
+@require_GET
```
Comment (eddieantonio) [inline CreeDictionary/CreeDictionary/views.py:106]: I learned about a new decorator today!

File: CreeDictionary/CreeDictionary/views.py [hunk 5]
```diff
@@ -86,18 +103,55 @@ def search_results(request, query_string: str):  # pragma: no cover
...
+    :raise 405 Method Not Allowed: when method other than GET is used
     """
-    lemma = Wordform.objects.get(id=lemma_id)
+    lemma_id = request.GET.get("lemma-id")
+    paradigm_size = request.GET.get("paradigm-size")
+    # guards
+    if lemma_id is None or paradigm_size is None:
+        return HttpResponseBadRequest("query params missing")
+    try:
+        assert float(lemma_id) == int(lemma_id)
```
Comment (eddieantonio) [inline CreeDictionary/CreeDictionary/views.py:126]: wat

File: CreeDictionary/CreeDictionary/views.py [hunk 6]
```diff
@@ -86,18 +103,55 @@ def search_results(request, query_string: str):  # pragma: no cover
...
+        lemma_id = int(lemma_id)
+    except (ValueError, AssertionError):
```
Comment (eddieantonio) [inline CreeDictionary/CreeDictionary/views.py:128]: So, even though I would never knowingly turn off assertions, the `assert` statement CAN be turned off (which is dumb, but still). See https://dbader.org/blog/python-assert-tutorial#:~:text=Caveat%20#1

I honestly have no idea what your `assert` is trying to do here.

File: CreeDictionary/tests/CreeDictionary_tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            ["-10", "FULL", HttpResponseBadRequest.status_code],
+            ["10", None, HttpResponseBadRequest.status_code],
+            ["5.2", "LINGUISTIC", HttpResponseBadRequest.status_code],
+            ["123", "LINUST", HttpResponseBadRequest.status_code],
+            [
+                "99999999",
+                "FULL",
+                HttpResponseNotFound.status_code,
+            ],  # we'll never have as many as 99999999 entries in the database so it's a non-existent id
+        ],
```
Comment (eddieantonio) [inline CreeDictionary/tests/CreeDictionary_tests/test_views.py:26]: 👍

File: CreeDictionary/tests/CreeDictionary_tests/test_views.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+        get_data: Dict[str, str] = {}
+        if lemma_id is not None:
+            get_data["lemma-id"] = lemma_id
+        if paradigm_size is not None:
+            get_data["paradigm-size"] = paradigm_size
+        response = c.get("/_lemma_details/", get_data)
+        assert response.status_code == expected_code
+
+    @pytest.mark.parametrize(("method",), (("post",), ("put",), ("delete",)))
```
Comment (eddieantonio) [inline CreeDictionary/tests/CreeDictionary_tests/test_views.py:41]: 👍

File: src/.eslintrc.yml [hunk 9]
```diff
@@ -26,4 +26,4 @@ rules:
     - never
   camelcase:
     - error
-  no-console: "warn"
+  no-console: ["warn", { allow: ["warn", "error"] }]
```
Comment (eddieantonio) [inline src/.eslintrc.yml:29]: Honestly, this is a good change 😅

File: src/css/styles.css [hunk 10]
```diff
@@ -624,7 +624,7 @@ html {
   border-spacing: 2em 0.5rem;
 }
 
-.paradigm__more-button {
+.paradigm__size-toggle-button {
```
Comment (eddieantonio) [inline src/css/styles.css:627]: This is a better name!

File: src/index.js [hunk 11]
```diff
@@ -59,6 +73,104 @@ document.addEventListener('DOMContentLoaded', () => {
...
+      if (typeof hash[1] !== 'undefined' && hash[1] !== null) {
+        url += '#' + hash[1]
+      }
+      return url
+    }
+    else {
+      return url
+    }
+  }
+}
```
Comment (eddieantonio) [inline src/index.js:128]: "ellemayo" did not write... the prettiest code.... but if it works? ¯\\\_(ツ)\_/¯

File: src/index.js [hunk 12]
```diff
@@ -59,6 +73,104 @@ document.addEventListener('DOMContentLoaded', () => {
...
+
+        } else {
+          paradigmFrag.querySelector('.paradigm__size-toggle-button-text').textContent = 'show more'
+          paradigmFrag.querySelector('.paradigm__size-toggle-plus-minus').textContent = '+ '
+        }
+        window.history.replaceState({}, document.title, updateQueryParam('paradigm-size', nextParadigmSize))
+        const oldParadigmNode = document.getElementById('paradigm')
+        oldParadigmNode.parentElement.replaceChild(paradigmFrag, oldParadigmNode)
+        paradigmSize = nextParadigmSize
+        setupParadigmSizeToggleButton() // prepare the new button
```
Comment (eddieantonio) [inline src/index.js:165]: Doesn't the current click event listener need to be removed first for this to not f things up?

File: CreeDictionary/CreeDictionary/views.py (hunk 6 above)
Comment (Madoshakalaka) [reply CreeDictionary/CreeDictionary/views.py:128]: this assert rule out the case when lemma-id is a float. if lemma-id is a float, without the assert. int() will cast the float to a integer.

Comment (eddieantonio) [reply CreeDictionary/CreeDictionary/views.py:128]: I tried it, and it just crashes :joy:

```python
>>> n = "3.14"
>>> float(n) == int(n)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: '3.14'
```

Please remove the assert!

Comment (Madoshakalaka) [reply CreeDictionary/CreeDictionary/views.py:128]: aaah somehow I recalled wrongly int will cast😔

File: CreeDictionary/CreeDictionary/views.py [hunk 13]
```diff
@@ -86,18 +103,54 @@ def search_results(request, query_string: str):  # pragma: no cover
...
     """
-    render paradigm table for a lemma
+    Render paradigm.html for a lemma. `index` view function renders a whole page that contains paradigm.html too.
+    This function, however, is used by javascript to dynamically replace the paradigm with the ones of different sizes.
+
+    `lemma-id` and `paradigm-size` are the expected query params in the request
+
+    4xx errors will have a single string as error message
+
+    :raise 400 Bad Request: when the query params are not as expected or inappropriate
```
Comment (Madoshakalaka) [inline CreeDictionary/CreeDictionary/views.py:116]: I hesitated between 400 and 422 Unprocessable Entity. Decided to use 400 cuz Django doesn't have a shortcut for 422

File: CreeDictionary/utils/enums.py [hunk 14]
```diff
@@ -11,13 +11,6 @@ class ParadigmSize(Enum):
     FULL = "FULL"
     LINGUISTIC = "LINGUISTIC"
 
-    @property
-    def display_form(self):
-        """
-        the form that we show to users on paradigm table
-        """
-        return self.value.capitalize()
-
```
Comment (Madoshakalaka) [inline CreeDictionary/utils/enums.py:20]: this is not used anywhere

File: cypress/integration/paradigm.spec.js [hunk 15]
```diff
@@ -1,53 +1,100 @@
-context('Paradigms', () => {
```
Comment (Madoshakalaka) [inline cypress/integration/paradigm.spec.js:1]: this `context` seems unneccary so I removed it

File: cypress/integration/paradigm.spec.js [hunk 16]
```diff
@@ -1,53 +1,100 @@
...
-          ctx = ctx.and('contain', wordform)
-        }
-      })
-    }
-
-    // TODO: the next test should be here, but it is broken because the
-    // upstream layouts are broken :/
-    it.skip('should display titles within the paradigm', () => {
-      // TODO: move this test into the previous test when the layout is fixed.
-      cy.visitSearch('minôsis')
```
Comment (Madoshakalaka) [inline cypress/integration/paradigm.spec.js:38]: I didn't touch any of these, dumb github diff, dumb

File: cypress/integration/paradigm.spec.js [hunk 17]
```diff
@@ -1,53 +1,100 @@
...
+    cy.get('[data-cy=paradigm]')
+      .as('paradigm')
+
+    // TODO: the layouts should be able to differentiate between titles and
+    // labels; currently, the specificiation is ambigous, hence, it's seen
+    // as a .paradigm-label, when it should be a .paradigm-title :/
+    cy.get('@paradigm')
+      .contains('.paradigm-title', 'Ownership')
   })
 })
```
Comment (Madoshakalaka) [inline cypress/integration/paradigm.spec.js:51]: dumb github diff, dumb

File: cypress/integration/paradigm.spec.js [hunk 18]
```diff
@@ -1,53 +1,100 @@
...
+    cy.get('[data-cy=paradigm]').contains('2p')
+
+    cy.get('[data-cy=paradigm-toggle-button]').click()
+    cy.wait(100)
+    // now are we back to basic?
+    cy.get('[data-cy=paradigm]').contains('Something is happening now')
+
+  })
+}
+)
```
Comment (Madoshakalaka) [inline cypress/integration/paradigm.spec.js:100]: new integration

File: cypress/support/commands.js [hunk 19]
```diff
@@ -60,10 +60,39 @@ Cypress.Commands.overwrite('visit', (originalVisit, url, options = {}) => {
...
+  })
+  queryParams = queryParams || {}
+  cy.visit(`/word/${encodeURIComponent(lemmaText)}/?${Object.entries(queryParams).map(([paramName, paramValue]) => paramName + '=' + encodeURIComponent(paramValue)).join('&')}`, {escapeComponents: false})
+  // test if a redirection happens
+  cy.location().should(
+    (loc)=>{
+      expect(loc.pathname, 'lemmaText and queryParams should be enough to disambiguate the lemma').to.eq(`/word/${encodeURIComponent(lemmaText)}/`)
+    }
+  )
+})
```
Comment (Madoshakalaka) [inline cypress/support/commands.js:98]: new command

File: src/index.js [hunk 20]
```diff
@@ -59,6 +83,114 @@ document.addEventListener('DOMContentLoaded', () => {
...
+ * read json data produced by django's `json_script` filter during HTML template generation
+ */
+function readDjangoJsonScript(id) {
+  const jsonScriptElement = document.getElementById(id)
+  if (jsonScriptElement){
+    return JSON.parse(jsonScriptElement.textContent)
+  }else{
+    return undefined
+  }
+}
```
Comment (Madoshakalaka) [inline src/index.js:96]: this

File: src/index.js (hunk 12 above)
Comment (eddieantonio) [reply src/index.js:185]: I did not notice

```javascript
        oldParadigmNode.parentElement.replaceChild(paradigmFrag, oldParadigmNode)
```

disregard my previous comment!

File: CreeDictionary/CreeDictionary/views.py (hunk 13 above)
Comment (eddieantonio) [reply CreeDictionary/CreeDictionary/views.py:116]: Haha, fair enough!

File: cypress/integration/paradigm.spec.js (hunk 16 above)
Comment (eddieantonio) [reply cypress/integration/paradigm.spec.js:38]: ¯\\\_(ツ)\_/¯

File: CreeDictionary/CreeDictionary/templates/CreeDictionary/base.html [hunk 21]
```diff
@@ -6,92 +6,109 @@
...
   <script src="{% url 'js_reverse' %}" type="text/javascript"></script>
   <script src={% static 'CreeDictionary/js/index.js' %}></script>
+
+  {% if lemma_id and paradigm_size %}
+    {% comment %}
+    meaning we are on lemma details/paradigm page. This variable can be utilized by javascript to dynamically request
+    other things (e.g. different sized paradigms) about the lemma
+    {% endcomment %}
+    {{ lemma_id|json_script:"lemma-id" }}
+    {{ paradigm_size.value|json_script:"paradigm-size" }}
```
Comment (eddieantonio) [inline CreeDictionary/CreeDictionary/templates/CreeDictionary/base.html:22]: 👍

File: CreeDictionary/CreeDictionary/views.py [hunk 22]
```diff
@@ -86,18 +103,54 @@ def search_results(request, query_string: str):  # pragma: no cover
...
+    else:
+        if lemma_id < 0:
+            return HttpResponseBadRequest(
+                "lemma-id is negative and should be non-negative"
+            )
+    try:
+        paradigm_size = ParadigmSize(paradigm_size.upper())
+    except ValueError:
+        return HttpResponseBadRequest(
+            f"paradigm-size is not one {[x.value for x in ParadigmSize]}"
```
Comment (eddieantonio) [inline CreeDictionary/CreeDictionary/views.py:138]: Error responses that tell you how to correct the error are 😍 😍 😍

File: cypress/integration/paradigm.spec.js [hunk 23]
```diff
@@ -1,53 +1,101 @@
...
+describe('paradigms are visitable from link', () => {
+  const lemmaText = 'niska'
+  it('shows basic paradigm', () => {
+    cy.visitLemma(lemmaText, {'analysis': 'niska+N+A+Sg'})
+    // Smaller/Lesser/Younger is an exclusive user friendly tag for BASIC paradigms
+    cy.get('[data-cy=paradigm]').contains('Smaller/Lesser/Younger')
+  })
+
+  it('shows full paradigm', () => {
+    cy.visitLemma(lemmaText, {'analysis': 'niska+N+A+Sg', 'paradigm-size': 'FULL'})
```
Comment (eddieantonio) [inline cypress/integration/paradigm.spec.js:63]: Oh, this is NICE!

File: cypress/integration/paradigm.spec.js [hunk 24]
```diff
@@ -1,53 +1,101 @@
...
+  it('shows linguistic paradigm', () => {
+    cy.visitLemma(lemmaText, {'analysis': 'niska+N+A+Sg', 'paradigm-size': 'LINGUISTIC'})
+    // DIMINUTIVE is an exclusive linguistic term for FULL paradigms
+    cy.get('[data-cy=paradigm]').contains('DIMINUTIVE')
+  })
+}
+)
+
+
+describe.only('paradigms can e toggled by the show more/less button', () => {
```
Comment (eddieantonio) [inline cypress/integration/paradigm.spec.js:77]:
```suggestion
describe('paradigms can e toggled by the show more/less button', () => {
```

File: CreeDictionary/CreeDictionary/templates/CreeDictionary/base.html [hunk 25]
```diff
@@ -6,92 +6,109 @@
...
-    {# The speaker-bio links template #}
-    <template id="template:speaker-bio-link">
-      <a href="" target="_blank" rel="noopener" data-cy="recordings-list__item-speaker">Learn more about <slot name="speaker-name"></slot>.</a>
-    </template>
-  {% endspaceless %}
+  {# The speaker-bio links template #}
+  <template id="template:speaker-bio-link">
+    <a href="" target="_blank" rel="noopener" data-cy="recordings-list__item-speaker">Learn more about
+      <slot name="speaker-name"></slot>
+      .</a>
```
Comment (eddieantonio) [inline CreeDictionary/CreeDictionary/templates/CreeDictionary/base.html:55]:
```suggestion
      <slot name="speaker-name"></slot>.
   </a>
```

File: src/index.js [hunk 26]
```diff
@@ -59,6 +83,114 @@ document.addEventListener('DOMContentLoaded', () => {
...
+/**
+ * read json data produced by django's `json_script` filter during HTML template generation
+ */
+function readDjangoJsonScript(id) {
+  const jsonScriptElement = document.getElementById(id)
+  if (jsonScriptElement){
+    return JSON.parse(jsonScriptElement.textContent)
+  }else{
+    return undefined
+  }
```
Comment (eddieantonio) [inline src/index.js:95]: Okay, so this is really dumb, but the following is equivillent:

```suggestion
  }
```

JavaScript functions return `undefined` if there is no return statement.

File: src/index.js [hunk 27]
```diff
@@ -59,6 +83,114 @@ document.addEventListener('DOMContentLoaded', () => {
...
+      text => {
+        const paradigmFrag = document.createRange().createContextualFragment(text)
+        if (allParadigmSizes.indexOf(nextParadigmSize) === allParadigmSizes.length - 1) {
+          paradigmFrag.querySelector('.paradigm__size-toggle-button-text').textContent = 'show less'
+          paradigmFrag.querySelector('.paradigm__size-toggle-plus-minus').textContent = '- '
+
+        } else {
+          paradigmFrag.querySelector('.paradigm__size-toggle-button-text').textContent = 'show more'
+          paradigmFrag.querySelector('.paradigm__size-toggle-plus-minus').textContent = '+ '
+        }
```
Comment (eddieantonio) [inline src/index.js:180]: I wish this was all handled on the server side or something, but I don't really have a good solution for that right now :/

Comment (Madoshakalaka) [reply src/index.js:180]: if you visit with e.g. /word/xxxx?paradigm-size=LINGUSTIC this is indeed rendered on the server side. so currently it's a mixed solution

File: src/index.js (hunk 26 above)
Comment (Madoshakalaka) [reply src/index.js:95]: nice. I actually didn't know
